src/core/occupation.jl

```diff
@@ -85,7 +91,14 @@ function find_occupation_fermi_metal(basis, energies, Psi)
     # taken with convergence criteria and the like
     εF = Roots.find_zero(εF -> compute_n_elec(εF) - n_electrons, (min_ε, max_ε),
                          Roots.Bisection())
-    @assert compute_n_elec(εF) ≈ n_electrons
+
+    n_electrons_integration = compute_n_elec(εF)
+    if abs(n_electrons_integration - n_electrons) > sqrt(eps(real(T)))
+        @warn("`find_occupation_gap_zero_temperature` has a mismatch in number " *
```

Comment:
wrong copy-paste here in the error message. Shouldn't these be asserts? It really shouldn't happen...

Reply:
Well but it does for some floating point types ... albeit the error is really neglible.

Reply:
Ah no, scratch that ... that's at the other instance, here it can be a proper assert.

Reply:
really ? The other one should really be fine also... that's in float16 I guess?

Reply:
No, it's for more accurate types like Double32 or Double64. I think the reason is because their density of representable floats is plainly different than for IEEE floats, such that `eps(T)` might not always be the best metric, or put another way: The prefactor for an `eps(T)` error bound might just be different, not sure.

Reply:
Seems a bit strange, the density thing shouldn't matter (we don't manipulate huge or tiny numbers). A more likely explanation is that something's converting to a less accurate precision under the radar

Reply:
Can you test if `filled_occ = T(2)` fixes it?

Reply:
That's also well possible. I'm not sure how `Double32` handles thinks which are not implemented generically ... could well fallback to `Float32` or so.

Reply:
Hm it probably doesn't because `2` is an int literal, and so not converted to Float64 before going into a double float... that's weird.

Reply:
Also `2` is exactly representable already for `Float16` :P

Reply:
Problem is likely `kweights`.
```
        kweights = [length(symops) for symops in ksymops]
        kweights = kweights / sum(kweights)
```
computes it in Float64. Since the eps for Float64s is about sqrt of the eps for Double64, there you go. This is hugely tricky to do systematically...

Reply:
Yes that could be it and that definitely should be changed!

Reply:
Places where we have to be careful. https://github.com/stevengj/ChangePrecision.jl is probably a good source of inspiration for things that can go wrong (as well as a good diagnostic tool). So apart from the obvious ones, things that create Float64 behind our backs:
- Irrationals
- Division/sqrt/exp/... of integers

Reply:
Hmm ... that gets me thinking: Since change precision has a way for changing default precision in `begin` ... `end` environments, we could use that for testing in the `DFTK.jl` module file sorrounding all includes. If our results are kind of agnostic to what we choose, we did a good job.

Reply:
Yeah that's what I meant by "diagnostic tool". So I did a quick grep for "\", things that might get us into trouble is pretty much all of PspHgh, SphericalHarmonics, bzmesh, `A_coeff(n) = (-1)^n / (factorial(n) * 4^n * sqrt(π))`

Reply:
Ah I see ... puh ... I can't be bothered to fix that all here. I made a separate issue for that, see #60.

Reply:
Hmm ... but some of that might not actually matter that much, because it's later assigned to arrays of the appropriate type.

Reply:
yeah but you lose precision, which matters if we want to push below Float64. Let's do that later, but keep the assert (which is pretty useful precisely because of this kind of things; in fact a better check would be to check against eps^2/3 or something)

Reply:
Ok, I dropped the relevant commit.